SourceKitStressTester/Sources/StressTester/SourceKitDocument.swift

```diff
@@ -471,6 +480,33 @@ class SourceKitDocument {
     }
   }
 
+  /// Send the `request` synchronously. Returns the response in case of success
+  /// and throws an error if the response didn't match the expected response of
+  /// `info` or if the request timed out.
+  private func sendRequestAndCheckResult(_ request: SourceKitdRequest, info: RequestInfo) throws -> SourceKitdResponse {
+    let (response, softError) = try sendWithTimeout(request, info: info)
```

Comment:
I haven't checked everything, but does anything actually call `sendWithTimeout` *without* running `throwIfInvalid`? Seems as though that function could just throw, in which case it could also throw the `softError` and you can avoid returning + checking it.

Reply:
Yeah, we do

https://github.com/apple/swift-stress-tester/blob/f09765337b3aa70ef248235b48c13c0bfa1c8aea/SourceKitStressTester/Sources/StressTester/SourceKitDocument.swift#L437

Reply:
Eh... I did see that one, but the only place it's used does end up doing a throw (see `codeComplete`). I'm not sure why we would ever not want to check really, it would be assuming the response didn't fail, which seems bad. Plus if we did want that we could just wrap it in a do/catch.

Reply:
After some offline discussion, I’ve changed `sendWithTimeout` to take a `validateResponse` closure, which is responsible for checking that the response contains the expected results. `sendWithTimeout` now also calls `throwIfInvalid` to check that the response is not an unexpected error response.

If `validateResponse` and `throwIfInvalid` pass but the requested exceeded half the timeout limit, `sendWithTimeout` will now throw a soft timeout error.